### DISCUSSION TITLE
Eloquent Collections should have a hasAny() method

### DISCUSSION THREAD

**Discussion by jackwh** [laravel/framework#42466] [Ideas]
[`Eloquent\Collection`](https://laravel.com/docs/9.x/eloquent-collections#available-methods) extends the base `Collection` class, making it easier to work with groups of models. But (unless I'm missing something, and I probably am), there's no "simple" way to determine if one Eloquent Collection has *any* of the same items as another.

It's definitely possible to do — but it's not as straightforward as it could be:

- Imagine a `User` model, which `BelongsToMany` related `Team` models
- We have two Users, and we want to find out if they work together in any of the same Teams
- With an Eloquent Collection, the most straightforward way to achieve this is by intersecting both sets:

```php
// Find the intersection of both teams, and if there are any intersected results, return true
return $user1->teams->intersect($user2->teams)->isNotEmpty();
```

The problem with intersecting is that it has to loop through *every* `Team` in both `User`'s collections. "Users and teams" is a trivial example, but for large datasets this is very inefficient. We just want to know if they have a single one in common — after the first match, the function can return immediately.

With a base Collection class, we could use the `hasAny()` method:

```php
$user1TeamIds = [1, 2, 3, 4];
$user2TeamIds = [3, 6, 9];

return collect($user1TeamIds)->hasAny($user2TeamIds); // returns true (after just 1 iteration)
```

But this isn't possible with `Eloquent\Collection`. It has several extended methods for working with Eloquent models, which can be wrangled to achieve the same end result (`contains`, `diff`, `except`, `intersect`, `unique`), but they all process the *entire result set*.

Of course, we could extract all keys from both collections, re-build a fresh base Collection with each, then perform the checks — but it's a shame to go the long-way round :) I'm happy to pop in a PR for this if I'm not missing anything obvious.

And a brief side note: only the [`has()`](https://laravel.com/docs/9.x/collections#method-has) collection method is documented, for some reason `hasAny()` isn't. Not sure why, maybe just an oversight?

**Comment by wunc** [comment]
Two points of clarification:

1. The methods `has()` and `hasAny()` check keys not values. And the latter is indeed [documented](https://laravel.com/docs/9.x/collections#method-hasany) (maybe it wasn't at the time of your comment).
2. Eloquent Collections [extend the base Collection](https://github.com/laravel/framework/blob/9.x/src/Illuminate/Database/Eloquent/Collection.php#L18). So you should be able to use both of those methods on an Eloquent Collection.

However, keeping to the point, if you want to stick with the Laravel nomenclature for checking values, then you probably want something named `containsAny()`. Spatie has a [Laravel Collection Macros package](https://github.com/spatie/laravel-collection-macros) with [exactly that](https://github.com/spatie/laravel-collection-macros#containsAny) (although looking at their code, they do use a intersect-then-count implementation, which as you mention is less efficient for larger data sets). Not sure if it works for Eloquent Collections, but I'm guessing it would.

Still, I agree that it would be nice to have that method in the core framework.
